### PR TITLE
Update DG

### PR DESCRIPTION
**Pull request by yiwen101**

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+The second stage is to parse each Parameter and flags. This work is delegated to the `typeParsingUtil` class, which is a utility class that contains many static methods that are responsible for parsing different types of user input that is reused across different command parsers.
+
+Then, the `EditPersonCommandParser` class is responsible for combining the results of the previous stage and construct the `EditPersonCommand` object.
+
+Finally, the `EditPersonCommand` class is responsible for executing the command and updating the model accordingly.
+
+The following section will discuss the implementation of the later three groups of classes in more detail.
+
+##### 1. ListEntry and ListEntryField class
+In TutorMates, Lesson and Person extends `ListEntry` class as they are displayed as items of the ___STUDENTS list___, ___SCHEDULE list___ respectively.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: TutorMates -> TutorMate

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
       * Many commands for user to remember, although they have similar ideas (e.g. `addPerson`, `addLesson` both involving adding a new object, which the user could have used a generic `add` method.)
       * Commands have to manually call the Ui to specify which Ui panels to show.
 
+### EditPersonCommand and Parser
+
+#### Background
+To boost usability and support more functionalities, many commands in TutorMate support multiple forms of valid user input.
+
+For example, when editing a person, the "INDEX" parameter can be conditionally omitted when editing the currently shown entry, the "edit" command word is overloaded for both "editLesson" and "editPerson" command, and for each of these two commands, many optional flags are available.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: Possibly reformat the parameter and command word to the code segments.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+The implementation of the `EditPersonCommand` and `EditPersonCommandParser` can serve as a good example to represent how we tackle this challenge at command and parser class level to achieve the desired flexibility and usability while keeping the code clean and maintainable.
+
+#### Implementation
+The parsing and execution process of the `EditPersonCommand` is achieved via the combination of Four groups of classes: `AddressBookParser`, `TypeParsingUtil`, `EditPersonCommandParser` and `EditPersonCommand`, each of which is responsible for a different stage of the parsing process. 
+
+Given a user input that is intended to be parsed into an `EditPersonCommand`, a high level description of the parsing process is as follows:
+
+The first stage is to understand that user intends to invoke the `EditPersonCommand` and delegate the parsing to the specialised parser class, `EditPersonCommmandParser`. This step is done by the `AddressBookParser` class, which behave like a "simple factory" via a giant "switch" statement. We will not discuss it in depth here as it is not very interesting.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: behave -> behaves
I think can remove the "as it is not very interesting part".

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+The second stage is to parse each Parameter and flags. This work is delegated to the `typeParsingUtil` class, which is a utility class that contains many static methods that are responsible for parsing different types of user input that is reused across different command parsers.
+
+Then, the `EditPersonCommandParser` class is responsible for combining the results of the previous stage and construct the `EditPersonCommand` object.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: construct -> constructing

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+Finally, the `EditPersonCommand` class is responsible for executing the command and updating the model accordingly.
+
+The following section will discuss the implementation of the later three groups of classes in more detail.
+
+##### 1. ListEntry and ListEntryField class
+In TutorMates, Lesson and Person extends `ListEntry` class as they are displayed as items of the ___STUDENTS list___, ___SCHEDULE list___ respectively.
+
+Each `ListEntry` object contains a list of `ListEntryField` objects, which are the fields of the `ListEntry` object. For example, a `Person` object contains a list of `ListEntryField` objects, which are the fields of the `Person` object, such as name, phone number, email address, etc.
+
+These two classes are used in multiple generic class and methods to achieve the flexibility and robustness of the parsing process.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: class -> classes

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+##### 2. Parsing of Flags and Parameters
+The `TypeParsingUtil` class parse all kinds of flags (which are all `ListEntryField` ) via a single powerful generic method, `parseField`. The method signature is as follows:
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: parse -> parses

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+The second parameter, `input`, is the user input that is intended to be parsed into the flag.
+
+The third parameter, `of`, is a self-defined functional interface that is to be used as the "factory" to create the flag object. The `of` method signature is as follows:
+```
+public interface Of<T extends ListEntryField> {
+    T apply(String str) throws IllegalArgumentException, ParseException;
+}
+```
+The last parameter, `isOptional`, is a boolean flag that indicates whether the flag is optional or not. If the flag is not found in the user input, this method will throw a `ParseException` if the flag is not optional, but will return a `null` object if the flag is optional. It is overloaded to has a default value of `true` when omitted.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: has -> have

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+// Parse the name flag, which is compulsory
+Name name = TypeParsingUtil.parseField("name", args, Name::of, false);
+
+// Parse the phone flag, which is optional
+Phone phone = TypeParsingUtil.parseField("phone", args, Phone::of);
+```
+##### 3. Parsing of the EditPersonCommand
+The `EditPersonCommandParser` class will parse the user input into an `EditPersonCommand` object. 
+
+This is achieved by combining the results of parsing flags in the previous stage and construct a `Person` object that contains the updated fields.
```
Comment (richiehx) [inline docs/DeveloperGuide.md:364]: construct -> constructing

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+Each of the `ListEntryField` has a global and static singleton default instance. For example, the default `Phone` is a `Phone` object with message "To be added."Please also note that it is not possible to create a `Phone` object with such message via public constructor and factory method.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: I think can add a space between "To be added." and Please

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+            person.setEmailIfNotDefault(parseField("email", args, Email::of));
+            person.setAddressIfNotDefault(parseField("address", args, Address::of));
+            person.setSubjectsIfNotDefault(parseField("subject", args, Subjects::of));
+            person.setTagsIfNotDefault(parseField("tag", args, Tags::of));
+            person.setRemarkIfNotDefault(parseField("remark", args, Remark::of));
+```
+##### 4. EditPersonCommand Implementation
+The `EditPersonCommand` class will execute the command and update the model accordingly.
+
+Realising the commonality between all kinds of edit commands, we have created an abstract generic class, `EditCommand`, to serve as template class containing the logic of editing a `ListEntry` object.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: should this be `AbstractEditCommand` instead of just `EditCommand`

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+The signature of the `EditCommand` class is as follows:
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: same as above

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+        deleteMethod = model::deletePersonForEdit;
+        addMethod = model::addPerson;
+        getClashingEntry = model::getPersonClashWith;
+        showMethod = model::showPerson;
+    }
+```
+
+#### Design considerations:
+
+**Aspect: How to keep the code clean and maintainable while providing the flexibility and variety of valid user input while .
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: I think the last while is extra

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+* **Alternative 1 (current choice):** Group the parsing and execution of a command into four groups of classes: `AddressBookParser`, `TypeParsingUtil`, `EditPersonCommandParser` and `EditPersonCommand`.
+    * Pros:
+        * Clear separation of concerns.
+        * Easy to maintain, easy to extend as for each new `ListEntryField` and `ListEntry` class that is added and fulfill the contract of the `ListEntryField` and `ListEntry` class, the parsing and execution of the command will almost be automatically supported.
+        * Good DRY principle. Only one method is needed to parse all kinds of flags, the knowledge of a given type of processing is encapsulated in a single place.
+    * Cons: Less flexible. It is not easy to add flags that has special parsing logic not supported by the current template.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: has -> have

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+
+* **Alternative 2:** Having even stricter constraints on the `ListEntryField` and `ListEntry` class, so that we can use reflection to automatically parse the flags.
+    * Pros: Potentially better extensibility.
+    * Cons:
+        * Difficult to implement.
+        * Even more difficult to debug.
+        * Even less flexible.
+        * Not easy to get the correct error message, as reelection call will through a `InvocationTargetException` instead of a `ParseException` when the parsing fails.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: what is a reelection call
i think through is supposed to be throw?

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
...
+**Aspect: How to keep the code clean and maintainable while providing the flexibility and variety of valid user input while .
+
+* **Alternative 1 (current choice):** Group the parsing and execution of a command into four groups of classes: `AddressBookParser`, `TypeParsingUtil`, `EditPersonCommandParser` and `EditPersonCommand`.
+    * Pros:
+        * Clear separation of concerns.
+        * Easy to maintain, easy to extend as for each new `ListEntryField` and `ListEntry` class that is added and fulfill the contract of the `ListEntryField` and `ListEntry` class, the parsing and execution of the command will almost be automatically supported.
+        * Good DRY principle. Only one method is needed to parse all kinds of flags, the knowledge of a given type of processing is encapsulated in a single place.
+    * Cons: Less flexible. It is not easy to add flags that has special parsing logic not supported by the current template.
+
+* **Alternative 2:** Having even stricter constraints on the `ListEntryField` and `ListEntry` class, so that we can use reflection to automatically parse the flags.
```
Comment (richiehx) [inline docs/DeveloperGuide.md]: what is reflection here

**Review by richiehx — COMMENTED** [review]
LGTM, just some nits

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 7 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md:364]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 11 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Fixed in the latest commit.

File: docs/DeveloperGuide.md (hunk 14 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: Thank you for pointing that out. Rephrased in the latest PR

File: docs/DeveloperGuide.md (hunk 15 above)
Comment (yiwen101) [reply docs/DeveloperGuide.md]: It is a feature in Java. Changed the elaboration in latest commit

**Review by pzl111 — APPROVED** [review]
LGTM

**Pull request merged**